### ISSUE TITLE
Graph writing speeds

### ISSUE THREAD

**Issue by jteresco** [TravelMapping/DataProcessing#626]
Graph writing appears to have slowed significantly on FreeBSD with the changes from #623 

With the current C++ siteupdate (5/21/2024), 8 threads, on FreeBSD, times for graph generation (computation and file writes):

Writing to the `/fast` SSD: ~86 sec
Writing to the `/home` magnetic drive: ~98 sec
Writing to `/ramdisk` RAM disk: ~96 sec

(only one run each for now)

But it doesn't seem to matter much where I write.  It also looks like the 8 threads are all pegging their cores during the graph generation process.

**Comment by yakra** [comment]
Oh boy. All this after I claimed it closed #518.
Mind if I log into noreaster & see what I can see?

**Comment by yakra** [comment]
https://travelmapping.net/updates/2024-05-21@15:12:31/logs/siteupdate.log
looks like to one done right before the OP here.
> DataProcessing @ 2cc40e097919a510f1c800ff7c14842e2f68c8d1

checks out; indeed the most recent revision.

WRT the approximate times, I use the following timestamps when measuring graph speed:
```
[14.8] Writing master TM graph files.
[107.1] Clearing HighwayGraph contents from memory.
```
so I get 92.3 s.

**Comment by yakra** [comment]
Compiled executables for 32829211cd8f93bb26a76d8ac339e4c454168a55 (AKA 52ef723da83485741366bcf6afdefe448dc49e69) & 2cc40e097919a510f1c800ff7c14842e2f68c8d1,
`siteupdate_3282` and `siteupdate_2cc4`.

Running
`/home/yakra/ytm/DataProcessing/siteupdate/cplusplus/check10.sh -h NEer -n -tt 8 -tp 4 -p 2 -r Graphs _3282 _2cc4`
and the same thing *(minus the `-h NEer` bit)* concurrently on bsdlab.
When done I'll do the same adding the `--null` option, which writes to /dev/null.

**Comment by yakra** [comment]
Reproducing these results so far.
I suppose it's *possible* that... the `std::ofstream` class can detect when it has a file in `/dev/null` "open", and skip the formatting of inserted doubles *(or \*anything\* really)* as a time-saver?

And maybe my [writing to /dev/null](https://github.com/TravelMapping/DataProcessing/pull/623#issuecomment-2119599132) isn't as representative of the real world as I'd been thinking all this time?

I'm going to fire up lab2 (CentOS) and see if I can reproduce the same results on Linux too. Heck, why not lab4 (Ubuntu) to o while I'm at it...

---

This could potentially also have impacts on userlogs & the DB.

**Comment by jteresco** [comment]
Thanks for checking into it.

Production site update was launched at 10:09 PM EDT, any benchmarking right around then could have been competing with it.

**Comment by yakra** [comment]
I. Can. Not. Believe this.
A wise man once said, *"There are lies, damn lies, and benchmarks."*

The columns of data below are:
* starting timestamp
* ending timestamp
* difference between the two
* total run time

---

### noreaster (FreeBSD 13.2)

```
[yakra@noreaster ~/ytm/DataProcessing/siteupdate/cplusplus]$ ./check10.sh -h NEer -n -tt 8 -tp 4 -p 2 -r Graphs _3282 _2cc4
NEer: siteupdate_3282, 8 thread(s)
13.4674	66.8072	 53.3398	70.9823	eta=Wed May 22 21:38:20 EDT 2024        
14.9634	70.2248	 55.2614	74.6430	eta=Wed May 22 21:38:30 EDT 2024        
siteupdate_3282, 8 thread(s), "master TM" -> "Clearing HighwayGraph contents from memory" average: 54.3006

NEer: siteupdate_2cc4, 8 thread(s)
14.0067	119.0338	 105.0271	119.6130	eta=Wed May 22 21:39:32 EDT 2024        
16.1430	125.0866	 108.9436	125.6047	eta=Wed May 22 21:40:09 EDT 2024        
siteupdate_2cc4, 8 thread(s), "master TM" -> "Clearing HighwayGraph contents from memory" average: 106.9853
```
```
[yakra@noreaster ~/ytm/DataProcessing/siteupdate/cplusplus]$ ./check10.sh --null -h NEer -n -tt 8 -tp 4 -p 2 -r Graphs _3282 _2cc4
NEer: siteupdate_3282, 8 thread(s)
12.7050	60.8581	 48.1531	61.3799	eta=Wed May 22 21:45:00 EDT 2024        
12.6330	62.4259	 49.7929	62.9824	eta=Wed May 22 21:45:04 EDT 2024        
siteupdate_3282, 8 thread(s), "master TM" -> "Clearing HighwayGraph contents from memory" average: 48.9730

NEer: siteupdate_2cc4, 8 thread(s)
9.8210	10.8307	 1.0097	11.3764	eta=Wed May 22 21:43:56 EDT 2024        
9.7894	10.7637	 .9743	11.3277	eta=Wed May 22 21:43:22 EDT 2024        
siteupdate_2cc4, 8 thread(s), "master TM" -> "Clearing HighwayGraph contents from memory" average: .9920
```

---

### bsdlab (FreeBSD 13.0)

```
[yakra@bsdlab ~/ytm/DataProcessing/siteupdate/cplusplus]$ ./check10.sh -n -tt 8 -tp 4 -p 2 -r Graphs _3282 _2cc4
bsdlab: siteupdate_3282, 8 thread(s)
20.4094	102.5987	 82.1893	108.7002	eta=Thu May 23 01:41:02 EDT 2024        
20.1705	107.7946	 87.6241	114.3095	eta=Thu May 23 01:41:12 EDT 2024        
siteupdate_3282, 8 thread(s), "master TM" -> "Clearing HighwayGraph contents from memory" average: 84.9067

bsdlab: siteupdate_2cc4, 8 thread(s)
21.2836	198.0550	 176.7714	198.9511	eta=Thu May 23 01:43:07 EDT 2024        
21.1611	197.4921	 176.3310	198.3914	eta=Thu May 23 01:44:05 EDT 2024        
siteupdate_2cc4, 8 thread(s), "master TM" -> "Clearing HighwayGraph contents from memory" average: 176.5512
```
```
[yakra@bsdlab ~/ytm/DataProcessing/siteupdate/cplusplus]$ ./check10.sh --null -n -tt 8 -tp 4 -p 2 -r Graphs _3282 _2cc4
bsdlab: siteupdate_3282, 8 thread(s)
19.1802	97.5847	 78.4045	99.2670	eta=Thu May 23 01:52:46 EDT 2024        
18.6800	103.1613	 84.4813	103.8550	eta=Thu May 23 01:52:54 EDT 2024        
siteupdate_3282, 8 thread(s), "master TM" -> "Clearing HighwayGraph contents from memory" average: 81.4429

bsdlab: siteupdate_2cc4, 8 thread(s)
14.0733	15.2657	 1.1924	16.0555	eta=Thu May 23 01:50:59 EDT 2024        
14.0775	15.3008	 1.2233	16.2240	eta=Thu May 23 01:50:03 EDT 2024        
siteupdate_2cc4, 8 thread(s), "master TM" -> "Clearing HighwayGraph contents from memory" average: 1.2078
```

---

## lab2 (CentOS 7)

```
[dan@lab2 cplusplus]$ ./check10.sh -n -tt 8 -tp 4 -p 2 -r Graphs _3282 _2cc4
lab2: siteupdate_3282, 8 thread(s)
113.3493	117.5240	 4.1747	118.5340	eta=Wed May 22 21:55:48 EDT 2024        
12.1756	16.3394	 4.1638	17.3561	eta=Wed May 22 21:52:24 EDT 2024        
siteupdate_3282, 8 thread(s), "master TM" -> "Clearing HighwayGraph contents from memory" average: 4.1692

lab2: siteupdate_2cc4, 8 thread(s)
12.1728	19.2653	 7.0925	20.2524	eta=Wed May 22 21:51:20 EDT 2024        
12.3656	19.4442	 7.0786	20.4375	eta=Wed May 22 21:50:49 EDT 2024        
siteupdate_2cc4, 8 thread(s), "master TM" -> "Clearing HighwayGraph contents from memory" average: 7.0855
```
```
[dan@lab2 cplusplus]$ ./check10.sh --null -n -tt 8 -tp 4 -p 2 -r Graphs _3282 _2cc4
lab2: siteupdate_3282, 8 thread(s)
11.9806	15.0405	 3.0599	16.0279	eta=Wed May 22 21:52:46 EDT 2024        
11.7957	14.8574	 3.0617	15.8427	eta=Wed May 22 21:52:44 EDT 2024        
siteupdate_3282, 8 thread(s), "master TM" -> "Clearing HighwayGraph contents from memory" average: 3.0608

lab2: siteupdate_2cc4, 8 thread(s)
11.7817	12.8166	 1.0349	13.8052	eta=Wed May 22 21:52:40 EDT 2024        
11.7108	12.7324	 1.0216	13.7719	eta=Wed May 22 21:52:40 EDT 2024        
siteupdate_2cc4, 8 thread(s), "master TM" -> "Clearing HighwayGraph contents from memory" average: 1.0282
```

---

## lab4 (Ubuntu 18.04)

```
yakra@lab4:~/ytm/DataProcessing/siteupdate/cplusplus$ ./check10.sh -n -tt 8 -tp 4 -p 2 -r Graphs _3282 _2cc4
lab4: siteupdate_3282, 8 thread(s)
64.4321	69.7670	 5.3349	70.9781	eta=Wed May 22 21:58:05 EDT 2024        
12.8422	18.3740	 5.5318	19.3759	eta=Wed May 22 21:56:21 EDT 2024        
siteupdate_3282, 8 thread(s), "master TM" -> "Clearing HighwayGraph contents from memory" average: 5.4333

lab4: siteupdate_2cc4, 8 thread(s)
12.7948	21.6652	 8.8704	22.8652	eta=Wed May 22 21:55:51 EDT 2024        
13.9076	23.2477	 9.3401	24.5730	eta=Wed May 22 21:55:39 EDT 2024        
siteupdate_2cc4, 8 thread(s), "master TM" -> "Clearing HighwayGraph contents from memory" average: 9.1052
```
```
yakra@lab4:~/ytm/DataProcessing/siteupdate/cplusplus$ ./check10.sh --null -n -tt 8 -tp 4 -p 2 -r Graphs _3282 _2cc4
lab4: siteupdate_3282, 8 thread(s)
12.2735	17.0719	 4.7984	18.6462	eta=Wed May 22 22:01:02 EDT 2024        
11.5002	15.6651	 4.1649	17.2744	eta=Wed May 22 22:01:02 EDT 2024        
siteupdate_3282, 8 thread(s), "master TM" -> "Clearing HighwayGraph contents from memory" average: 4.4816

lab4: siteupdate_2cc4, 8 thread(s)
12.1213	13.6509	 1.5296	14.7393	eta=Wed May 22 22:00:56 EDT 2024        
11.5833	13.1592	 1.5759	14.1159	eta=Wed May 22 22:00:54 EDT 2024        
siteupdate_2cc4, 8 thread(s), "master TM" -> "Clearing HighwayGraph contents from memory" average: 1.5527
```

**Comment by yakra** [comment]
> This could potentially also have impacts on userlogs & the DB.

No need for an additional benchmark run; I can extract these data from the siteupdate logs produced by the one I just did using [table.sh](https://github.com/yakra/tmtools/blob/master/shell_scripts/table.sh).

**Comment by yakra** [comment]
**Writing userlogs to /dev/null @ 8 threads**
| | 3282921 | 2cc40e0 |
| -- | -- | -- |
noreaster | 3.1203 | .2911
bsdlab | 5.4446 | .7018
lab2 | .6270 | .5986
lab4 | .8863 | .8364

**Writing userlogs to disk @ 8 threads**
| | 3282921 | 2cc40e0 |
| -- | -- | -- |
noreaster | TBD | TBD
bsdlab | 5.8518 | 6.8314
lab2 | .6504 | .7211
lab4 | .9351 | .9449

---

> No need for an additional benchmark run; I can extract these data from the siteupdate logs produced by the one I just did using [table.sh](https://github.com/yakra/tmtools/blob/master/shell_scripts/table.sh).

Whoops, not quite. The 2nd benchmark run (to /dev/null) overwrote the first, more relevant, logs.
Let me know when it's safe to do a run on noreaster again.

**Comment by yakra** [comment]
This is just a more readable regurgitation of the post [above](https://github.com/TravelMapping/DataProcessing/issues/626#issuecomment-2126065263).

**Writing graphs to /dev/null @ 8 threads** (lies, damn lies, and benchmarks)
| | 3282921 | 2cc40e0 |
| -- | -- | -- |
noreaster | 48.9730 | .9920
bsdlab | 81.4429 | 1.2078
lab2 | 3.0608 | 1.0282
lab4 | 4.4816 | 1.5527

**Writing graphs to disk @ 8 threads** (real world)
| | 3282921 | 2cc40e0 |
| -- | -- | -- |
noreaster | 54.3006 | 106.9853
bsdlab | 84.9067 | 176.5512
lab2 | 4.1692 | 7.0855
lab4 | 5.4333 | 9.1052

**Comment by jteresco** [comment]
Wow, a lot happened while I was looking away.  You are welcome to run on noreaster any time.  I don't have a group working on METAL this summer so the machine is mostly bored and would probably be excited to have some more benchmarks to run...

**Comment by yakra** [comment]
I just [built & installed fmt](https://fmt.dev/latest/usage.html) on BiggaTomato seconds before you posted. I've yet to try it out.
Had to build & install cmake itself (version 3.15.5) but thankfully that process was painless after doing it on lab3.
noreaster doesn't have cmake installed; looks like pkg can do it. (Version 3.26.1, looks like?)

**Comment by jteresco** [comment]
cmake installed on noreaster.

**Comment by yakra** [comment]
fmt is built & installed on bsdlab. Here's the process I used:
```sh
# get https://github.com/fmtlib/fmt/releases/download/10.2.0/fmt-10.2.0.zip
# however you choose; I SCPed it over from my desktop
su
unzip fmt-10.2.0.zip
cd fmt-10.2.0
cmake -DBUILD_SHARED_LIBS=TRUE .
# OK, I didn't actually do the -DBUILD_SHARED_LIBS=TRUE bit. Might have another
# go soon, and do that part. I haven't successfully used the shared libs on
# BiggaTomato, but it might be nice to have them available in the future.
gmake -j 24 # or however many threads you want to use
make install

ln -s /usr/local/include/fmt /usr/include/fmt
```
That last step feels like a bit of a kluge, but I needed to do *something* to get clang to find the fmt header files, and that worked.
Don't know how familiar you are with cmake, or how to change the directory something installs to *(whether there's a fairly common way these `gmake install` type installations do things, or if it varies widely.)* of if there's a way to tell clang where to look for header files (other than on a per-project basis)...
If you know of something off the top of your head, I'd love to be enlightened. :grin:
If not, not to worry. I got something working at least.

Here's a quick little proof of concept:
```c++
#define FMT_HEADER_ONLY
#include <fmt/format.h>
#include <iostream>

int main()
{	char buf[20];
	*fmt::format_to(buf, "{}", 9235.1470421475879) = 0;
	std::string str = fmt::format("{}", 9235.1470421475879);
	std::cout << buf << std::endl;
	std::cout << str << std::endl;
}
```

**Comment by yakra** [comment]
Nobody's forks other than mine have been updated since this happened.
Do we...
* Reset master back to 52ef723da83485741366bcf6afdefe448dc49e69?
* Go with the "everything is part of the commit history, even the mistakes" philosophy?

**Comment by jteresco** [comment]
I am ok with either way to get back where we were on FreeBSD efficiency (reset or leave commits).  The bigger question to me is whether we would want some conditional compilation (`#ifdef` blocks) to use the old or new code depending on where it's being compiled.  That makes a messier codebase but we wouldn't lose what you have done in cases when it makes sense to use.

**Comment by yakra** [comment]
Based on the last sentence, I assume by "old or new code" you mean 52ef723da83485741366bcf6afdefe448dc49e69 (before #623) & 2cc40e097919a510f1c800ff7c14842e2f68c8d1 (after #625).

I don't see any need to keep the code introduced in those commits.
After re-running the old benchmark on BiggaTomato (post [above](https://github.com/TravelMapping/DataProcessing/issues/626#issuecomment-2126126103) edited), turns out the new code doesn't help real-world performance *anywhere.*

Another potential use case for conditional compilation:
Using the old code (or at least changing to `snprintf`) if the `fmt` headers are not available.
But I'd rather not do that either; I'd prefer to just make `fmt` a prerequisite. It's easy enough to install.

---

**Progress:**
Starting from 32829211cd8f93bb26a76d8ac339e4c454168a55 (which has no diff from 52ef723da83485741366bcf6afdefe448dc49e69), I've replaced `sprintf` calls in the 5 `.cpp ` files modified in 2349b1f1a5f75ec17d239495e637f3838fe09dc4 with calls to `fmt::format_to`.

**Noreaster:**
Assuming `fmt` isn't installed yet, I just plunked it down in `/home/yakra/ytm/DataProcessing/siteupdate/cplusplus/fmt-10.2.0/`, and `"include"`d the headers rather than `<include>` them.
**21.6 s** writing graphs **to disk** (magnetic) @ 8 threads.

**Format string compilation:**
I tried it out on noreaster.
**21.2 s** writing graphs to magnetic disk @ 8 threads.
Fairly small difference, but still significant, I believe. Consistently faster at any number of threads on bsdlab "writing" to /dev/null, taking about 91% as long on average.
That said, I'm reluctant to use this for production. We don't just need C++17 support, but also recent enough standard library headers, `<type_traits>` at the very least. That's no bueno for 5 of my 8 machines. I'd rather not upgrade clang again for a while. ;)

**ToDo:**
* [ ] Try out [std::ostream Support](https://fmt.dev/latest/api.html#ostream-api).
  This is a bit like [#623](https://github.com/TravelMapping/DataProcessing/pull/623) in that we format directly into the output file without messing around with an intermediate `char` array. May be faster?
* [ ] Try out `fmt` for integers too. It's supposedly faster than ostreams.

**Comment by jteresco** [comment]
I just installed the libfmt package on noreaster, so it should be available to any C++ compiler in the standard include and lib paths.

FYI, I am planning to do an OS update for noreaster soon, at which time I'll also update all of the installed packages to their latest versions.  I am just working to make sure I do so on a day when I can be at my office for several hours in case I break things to the point where I need physical access to the console.

**Comment by yakra** [comment]
Thanks for installing libfmt.
Looks like you've experienced the same issue I did on bsdab, that caused me to make the symbolic link...

from running `clang++ -v`:
```
#include <...> search starts here:
 /usr/include/c++/v1
 /usr/lib/clang/14.0.5/include
 /usr/include
```

libfmt got installed to `/usr/local/include/fmt`.

**Comment by yakra** [comment]
https://clang.llvm.org/docs/ClangCommandLineReference.html
Trying out the `isystem` flag in Makefile.

**Comment by jteresco** [comment]
I'm surprised `/usr/local/include` isn't part of the standard include path, but a `-I` or what looks to be a more modern `-isystem` parameter is certainly appropriate.

**Comment by yakra** [comment]
> **ToDo:**
> * [x] Try out [std::ostream Support](https://fmt.dev/latest/api.html#ostream-api).
>   This is a bit like [#623](https://github.com/TravelMapping/DataProcessing/pull/623) in that we format directly into the output file without messing around with an intermediate `char` array. May be faster?

No noticeable difference. Just noise in the data.

Another thing to try out *may* be `fmt::output_file`; [changing the buffer size can potentially improve performance](https://vitaut.net/posts/2020/optimal-file-buffer-size/).
Can this be done with a regular C++ `std::oftsream`? Yes. I tried out a quick-n-dirty (full of memory leaks)
> tmgfile.[rdbuf()](https://cplusplus.com/reference/ios/ios/rdbuf/)->[pubsetbuf](https://cplusplus.com/reference/streambuf/streambuf/pubsetbuf/)(new char[8192], 8192);

on noreaster. Again, no noticeable difference.

It may be that `fmt::output_file` uses C `FILE`s instead of C++ `ofstream`s under the hood. I'm not 100% certain from looking at the code. Not sure how much the 2 would differ WRT buffer size; can't imagine much.
May try it out, but low priority.

**Comment by yakra** [comment]
> I am ok with either way to get back where we were on FreeBSD efficiency (reset or leave commits).

I've reset yakra/master to 52ef723da83485741366bcf6afdefe448dc49e69 and force pushed.
Shall we do the same with TM/master?

**Comment by yakra** [comment]
Sure, people's *forks* haven't been updated, but what about anyone who's *cloned* the repo, without keeping a copy on GitHub?
Let's skip the reset, and take the "everything is part of the commit history, even the mistakes" route.

Pull request later today.

**Comment by jteresco** [comment]
Sure, I'm good with some mistakes in the history.  And we know not to try it again.

**Comment by yakra** [comment]
> Pull request later today.

Scratch that. Found & fixed a last-minute bug, then did another benchmark round with the new binary.
Charts are *mostly* ready; need to take the time to do the full write-up.

**TLDR**, postponed.
I *could* have it by the end of tomorrow, but that may be a touch optimistic.

**Comment by jteresco** [comment]
This sounds promising, thanks as always for all of this!

**Comment by yakra** [comment]
Do you have `{fmt}` *(I think the braces are part of its "official" name)* installed on the Mac?

**Comment by yakra** [comment]
> FYI, I am planning to do an OS update for noreaster soon, at which time I'll also update all of the installed packages to their latest versions. I am just working to make sure I do so on a day when I can be at my office for several hours in case I break things to the point where I need physical access to the console.

Did this happen today?
Clang is generating pages of fatal errors due to not finding header files.
```sh
[yakra@noreaster ~/ytm/DataProcessing/siteupdate/cplusplus]$ gmake siteupdate
classes/GraphGeneration/HighwayGraphMT.d
In file included from classes/GraphGeneration/HighwayGraph.cpp:2:
In file included from classes/GraphGeneration/HighwayGraph.h:10:
In file included from /usr/include/c++/v1/list:201:
/usr/include/c++/v1/__algorithm/equal.h:20:10: fatal error: '__string/constexpr_c_functions.h' file not found
   20 | #include <__string/constexpr_c_functions.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```
```sh
[yakra@noreaster ~/ytm/DataProcessing/siteupdate/cplusplus]$ ls -l /usr/include/c++/v1/list
-r--r--r--  1 root  wheel  71793 May 29 15:17 /usr/include/c++/v1/list
```

**Comment by rickmastfan67** [comment]
@yakra check your email.  There's an important email from Jim in it.

**Comment by jteresco** [comment]
I am sure clang got updated, and I just removed and reinstalled gcc13 in case that's helpful.  valgrind is also now on noreaster.

**Comment by jteresco** [comment]
Definitely is the case `clang++` on noreaster has been broken by the OS upgrade (I assume).  Looks like I'm not the only one who's seeing this problem, for better or for worse.

**Comment by yakra** [comment]
> I just removed and reinstalled gcc13 in case that's helpful.

I gave it a whirl for giggles. It helpfully asked if I meant to `#include <string>` in a few of my header files.
LOL, if it needs it and doesn't have it, how has clang been compiling it this whole time? I would think that clang should throw the same error and refuse to compile instead of just silently allowing me to write my code wrong. :3
But anyway!
As expected, siteupdate compiles but does not run:
`ld-elf.so.1: /usr/local/lib/gcc10/libstdc++.so.6: version GLIBCXX_3.4.32 required by ./siteupdate not found`
ISTR reading around when we first tried to get C++ siteupdate up & running on noreaster that ... glibc, glibcxx, whatever it was, didn't ship with FreeBSD because of incompatible licenses, and clang was made the default compiler for FreeBSD. ...Or something.
This only affected the threaded version, so I went ahead with parallel development of siteupdateST, which would at least run on FreeBSD when compiled with g++.
What I did *not* expect was for siteupdateST to fail to run, with the same error message.
Oy vey...
Meanwhile on bsdlab, both siteupdateST *and* siteupdate run just fine using g++. :shrug:

> valgrind is also now on noreaster.

Thanks! One of the versions I `valgrind`ed *(valground?)* on bsdlab was also compiled on noreaster before all this began, so I'm trying it out right now.

I definitely feel your pain on having a b0rked FreeBSD installation.
While mine *may* be back to normal, or mostly so, I don't like these [memory leaks](https://github.com/TravelMapping/DataProcessing/issues/627#issuecomment-2135730084) that valgrind is reporting. I have to wonder, if I did a fresh install, if I'd still see them...

**Comment by jteresco** [comment]
I cobbled together what might be a working `clang` on noreaster.

**Comment by yakra** [comment]
It compiled siteupdate @ 0b2982b8cf5f339f0288003941d55e7535089baa without errors, just a few warnings about unqualified calls to `move` instead of `std::move`. (Another case of "How does this actually compile?")
`compare_all.sh` against 32829211cd8f93bb26a76d8ac339e4c454168a55 looks good.
Valgrind is in progress. Currently writing continent graphs, but this takes a while.

So I'm deadly curious, just what did this cobbling together entail?

**Comment by jteresco** [comment]
It actually wasn't that bad.  Some files under `'/usr/include/c++` that were previously files have become directories, and the update process had a bug that didn't deal with this.  So in the end (after a lot of investigation and thinking about best ways to resolve), I grabbed the whole FreeBSD source code distribution and just put everything that's supposed to be there, in there.  There was one extra step at the end to copy back in one config file from the original.

**Comment by yakra** [comment]
Wow. Sounds like a lot of detective work. Congrats!
